tests: kernel: timer_api: more attempts to make this pass on every Nordic board instance

pabigot opened the pull request:
It's known that Nordic boards use a different clock source for the system clock and the CPU (which controls the duration of `k_busy_wait()`.  The skew between these clocks is dependent on board electronics, temperature, and who knows what else.  However, the test was originally written presuming these clocks were in sync, which is probably true for most other platforms.

This PR adds another way of estimating the maximum potential error that can be observed on a Nordic board running this test, and uses this in two test cases that have been observed to fail (many of the others have not, but it's probably just a matter of time).

I'm not entirely comfortable with merging this at this point in the release cycle, because I can't guarantee it won't have an effect on other boards (most likely making them pass when they shouldn't). ~For one thing it may be that the correct default slew for non-Nordic boards is 0 ppm, but I've made it 1 ppm just so it's less likely other boards will fail when they shouldn't.~ updated version makes this 0.

For another if we're going to try to do this sort of error estimation we should do it outside this test and make it dependent on actual clock configurations.  That's something we might consider for 2.4.0.

pabigot committed:
tests: kernel: timer_api: use slew for remaining timer threshold

This test sets a timer using one clock, waits using a second clock,
then sees whether the remaining time is the expected value.  When the
two clocks are skewed the comparison requires a threshold.  Provide a
means to estimate the maximum expected error.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
tests: kernel: timer_api: use slew for duration timer test

Reduce the duration of the timer test so that it will fire before the
busywait elapses even in the worst case of slew between the tick and
busy-wait clocks.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

carlescufi commented:
@andyross I'd very much appreciate your take on this change.

pabigot commented:
If we aren't going to merge this proposal for 2.3.0 to make Nordic boards pass, I would prefer to solve this another way.  Possibly by measuring the actual slew before the test starts up, or reworking it to support a reference clock other than `k_busy_wait()`.

andyross commented:
I'm not opposed to this technique, really.  Fudge factors for other areas exist in the test suite already, this really isn't any different.

I do think it's an awful lot of complexity to be swallowing in the test for what I continue to believe shouldn't be a real problem in practice.  We have an API that nRF can't quite implement in the default configuration (different API calls that take realtime units that "should" match, but that are in practice skewed enough to confuse test code).  But the test requirements, while they're not wrong per the API docs, are decidedly uncommon in production code.  It's very reasonable that nRF can get away with just documenting this and telling users not to spin so long.

But nRF **does** have a mode that meets the requirements!  Why not just turn on the crystal for this one test?  Or even this one test case?  Personally I'd rather see the complexity bandwidth spent on coming up with a clean API for configuring the board oscillator in the test than building a rig around fudge factors.

But frankly at this stage in the release that's not worth a -1 either.  This isn't so bad, we have weird stuff in the tests already.

pabigot committed:
tests: kernel: timer_api: compensate for fast clocks

When HFCLK has a slew making it faster than LFCLK the busy wait can
expire before the timer fires.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
tests: kernel: timer_api: compensate inaccurate conversion

When millisecond/tick conversion is not exact tick delta's are
dependent on the initial tick value.  In those cases exact comparisons
need to also allow an adjacent value.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
Had to add two more evasions to make this pass on my nrf51dk.

carlescufi commented:
> But nRF **does** have a mode that meets the requirements! Why not just turn on the crystal for this one test? Or even this one test case? Personally I'd rather see the complexity bandwidth spent on coming up with a clean API for configuring the board oscillator in the test than building a rig around fudge factors.

The problem is that if we do that then we are testing something that no one actually uses in the real world for Nordic platforms. Or at least that was our conclusion the last time we discussed this with @nordic-krch  and @pabigot

pabigot commented:
The current version has passed over 1000 runs on these platforms:

* nrf9160dk_nrf9160
* nrf5340pdk_nrf5340_cpuapp
* nrf52840dk_nrf52840
* nrf52dk_nrf52832
* nrf51dk_nrf51422

which may be enough to remove the hardware block.  If not need testing by @ioannisg because that's all the variants I have.

While switching to a different clock configuration does not represent real-world use, neither does this test in its current form.  That it passes does not mean that similar code in applications will behave as expected, because it only passes because we accept "no" as an answer to questions like "did the timer fire", just because we can't confirm it.

A case can be made that having this test use a special clock configure is a better solution, because then we can remove all the fudge factors and verify that the implementation does the right thing **if** the clocks are synchronized.  I'm going to try that next.

ioannisg commented:
I can provide testing results for 52833as well, hang on

pabigot commented:
To try to make the clocks synchronous I added this to a test-specific Kconfig
```
choice CLOCK_CONTROL_NRF_SOURCE
	default CLOCK_CONTROL_NRF_K32SRC_SYNTH
endchoice
```
The version on master fails `test_timer_remaining` on my nrf51dk_pca10028  because the error is 4 ticks, larger than the already relaxed limit of 3 ticks.

I'm not sure why this is, though inspecting the `CLOCK` peripheral registers directly confirms that the application started and ended with LFSRC Synth and HFSRC RC.  And the `reel_board` that normally fails reliably passes reliably.

carlescufi commented:
@nordic-krch could you please review?
